server/api.go

```diff
@@ -45,12 +50,58 @@ type PRDetails struct {
 	Reviews            []*github.PullRequestReview `json:"reviews"`
 }
 
+type HTTPHandlerFuncWithUser func(w http.ResponseWriter, r *http.Request, userID string) http.HandlerFunc
+
 func writeAPIError(w http.ResponseWriter, err *APIErrorResponse) {
 	b, _ := json.Marshal(err)
 	w.WriteHeader(err.StatusCode)
 	w.Write(b)
 }
 
+func (p *Plugin) initialiseAPI() {
+	p.router = mux.NewRouter()
+
+	webHookRouter := p.router.PathPrefix(API_WEBHOOK).Subrouter()
+	oauthRouter := p.router.PathPrefix(API_OAUTH).Subrouter()
+	apiRouter := p.router.PathPrefix(API_PREFIX).Subrouter()
+
+	webHookRouter.HandleFunc("/", p.handleWebhook)
+
+	oauthRouter.HandleFunc("/connect", p.extractUserMiddleWare(p.connectUserToGitHub, false))
+	oauthRouter.HandleFunc("/complete", p.extractUserMiddleWare(p.completeConnectUserToGitHub, false))
+
+	apiRouter.HandleFunc("/connected", p.extractUserMiddleWare(p.getConnected, true)).Methods("GET")
+	apiRouter.HandleFunc("/todo", p.extractUserMiddleWare(p.postToDo, true))
+	apiRouter.HandleFunc("/reviews", p.extractUserMiddleWare(p.getReviews, false)).Methods("GET")
+	apiRouter.HandleFunc("/yourprs", p.extractUserMiddleWare(p.getYourPrs, false)).Methods("GET")
+	apiRouter.HandleFunc("/prsdetails", p.extractUserMiddleWare(p.getPrsDetails, false)).Methods("POST")
+	apiRouter.HandleFunc("/searchissues", p.extractUserMiddleWare(p.searchIssues, false)).Methods("GET")
+	apiRouter.HandleFunc("/yourassignments", p.extractUserMiddleWare(p.getYourAssignments, false)).Methods("GET")
+	apiRouter.HandleFunc("/createissuecomment", p.extractUserMiddleWare(p.createIssueComment, false)).Methods("POST")
+	apiRouter.HandleFunc("/mentions", p.extractUserMiddleWare(p.getMentions, false)).Methods("GET")
+	apiRouter.HandleFunc("/unreads", p.extractUserMiddleWare(p.getUnreads, false)).Methods("GET")
+	apiRouter.HandleFunc("/settings", p.extractUserMiddleWare(p.updateSettings, false))
+	apiRouter.HandleFunc("/user", p.extractUserMiddleWare(p.getGitHubUser, true)).Methods("POST")
+}
+
+func (p *Plugin) extractUserMiddleWare(handler HTTPHandlerFuncWithUser, jsonResponse bool) http.HandlerFunc {
+
+	return func(w http.ResponseWriter, r *http.Request) {
+		userID := r.Header.Get("Mattermost-User-ID")
+		if userID != "" {
+			handler(w, r, userID)(w, r)
+			return
+		}
+
+		if jsonResponse {
+			writeAPIError(w, &APIErrorResponse{ID: "", Message: "Not authorized.", StatusCode: http.StatusUnauthorized})
+		} else {
+			http.Error(w, "Not authorized", http.StatusUnauthorized)
+		}
```

Comment:
I'm 1/5 that the logic should be inverted given that https://github.com/mattermost/mattermost-plugin-github/pull/248/files#diff-0d5a44e9b96f048fa52b477ce6be6253R96-R100 is the error handling.